bson/primitive/primitive_test.go

```diff
@@ -199,3 +200,334 @@ func TestDateTime(t *testing.T) {
+			assert.Equal(t, want, got)
+		})
+	}
+}
+
+func TestUnmarshalingA(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []struct {
+		name string
+		test A
+	}{
+		{
+			"nil",
+			nil,
+		},
+		{
+			"empty",
+			A{},
+		},
+		{
+			"non-empty",
+			A{
+				"hello world",
+				3.14159,
+				true,
+				nil,
+				A{2.71828, "Lorem ipsum", nil, false, D{{"Lorem", "ipsum"}}},
+				D{{"foo", "bar"}},
+			},
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			b, _ := json.Marshal(tc.test)
+			var got A
+			err := json.Unmarshal(b, &got)
+			assert.NoError(t, err)
+			assert.Equal(t, tc.test, got)
+		})
+	}
+
+	for _, tc := range []struct {
+		name string
+		test string
+	}{
+		{
+			"illegal",
+			`nil`,
+		},
+		{
+			"invalid",
+			`[null, 3.142ipsum]`,
+		},
+		{
+			"malformatted",
+			`[null, 3.142,]`,
+		},
+		{
+			"truncated",
+			`[2.71828, "Lorem ipsum", null,`,
+		},
+		{
+			"object type",
+			`{"pi": 3.142}`,
+		},
+		{
+			"number type",
+			`3.142`,
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			var a []interface{}
+			want := json.Unmarshal([]byte(tc.test), &a)
+			var b A
+			got := json.Unmarshal([]byte(tc.test), &b)
+			switch w := want.(type) {
+			case *json.UnmarshalTypeError:
+				w.Type = reflect.TypeOf(b)
+				require.IsType(t, want, got)
+				g := got.(*json.UnmarshalTypeError)
+				assert.Equal(t, w, g)
+			default:
+				assert.Equal(t, want, got)
+			}
+		})
+	}
+}
+
+func TestUnmarshalingM(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []struct {
+		name string
+		test M
+	}{
+		{
+			"nil",
+			nil,
+		},
+		{
+			"empty",
+			M{},
+		},
+		{
+			"non-empty",
+			M{
+				"a": "Lorem ipsum",
+				"b": 3.142,
+				"c": true,
+				"d": nil,
+				"f": A{"Lorem ipsum", 3.142, true, nil, M{"Lorem": "ipsum"}},
+				"g": M{"foo": "bar"},
+			},
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			b, _ := json.Marshal(tc.test)
+			var got M
+			err := json.Unmarshal(b, &got)
+			assert.NoError(t, err)
+			assert.Equal(t, tc.test, got)
+		})
+	}
+
+	for _, tc := range []struct {
+		name string
+		test string
+	}{
+		{
+			"illegal",
+			`nil`,
+		},
+		{
+			"invalid",
+			`{"pi": 3.142ipsum}`,
+		},
+		{
+			"malformatted",
+			`{"pi", 3.142}`,
+		},
+		{
+			"truncated",
+			`{"pi": 3.142`,
+		},
+		{
+			"array type",
+			`["pi", 3.142]`,
+		},
+		{
+			"string type",
+			`"pi"`,
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			var a map[string]interface{}
+			want := json.Unmarshal([]byte(tc.test), &a)
+			var b M
+			got := json.Unmarshal([]byte(tc.test), &b)
+			switch w := want.(type) {
+			case *json.UnmarshalTypeError:
+				w.Type = reflect.TypeOf(b)
+				require.IsType(t, want, got)
+				g := got.(*json.UnmarshalTypeError)
+				assert.Equal(t, w, g)
+			default:
+				assert.Equal(t, want, got)
+			}
+		})
+	}
+}
+
+func TestUnmarshalingD(t *testing.T) {
```

Comment:
Optional: Consider a more descriptive test name (borrowed from testable examples name conventions).

```suggestion
func TestD_UnmarshalJSON(t *testing.T) {
```